### PR TITLE
[BEAM-646] Add Pipeline#replaceTransform

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
This is the base method for Pipeline Surgery. It takes a
PTransformMatcher and a PTransformOverrideFactory, and replaces all
matching PTransforms with the result of the Override Factory.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles 
CC: @amitsela

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 1]
```diff
@@ -394,6 +426,33 @@ public String toString() {
...
+          TransformT extends PTransform<? super InputT, OutputT>>
+      void applyReplacement(
+          Node original,
+          PTransformOverrideFactory<InputT, OutputT, TransformT> replacementFactory) {
+    // Names for top-level transforms have been assigned. Any new collisions are within a node
+    // and its replacement.
+    getOptions().setStableUniqueNames(CheckEnabled.OFF);
+    PTransform<InputT, OutputT> replacement =
+        replacementFactory.getReplacementTransform((TransformT) original.getTransform());
+    // if (replacement == original.getTransform()) { return; }
```
Comment (amitsela) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Comment in ?

Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 2]
```diff
@@ -167,6 +172,33 @@ public PBegin begin() {
...
+  public void replace(
+      final PTransformMatcher matcher, PTransformOverrideFactory replacementFactory) {
+    final Collection<Node> matches = new ArrayList<>();
+    transforms.visit(
+        new PipelineVisitor.Defaults() {
+          @Override
+          public CompositeBehavior enterCompositeTransform(Node node) {
+            if (!node.isRootNode() && matcher.matches(node.toAppliedPTransform())) {
+              matches.add(node);
+              // This node has been replaced. It should not be visited.
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: s/has/will be/

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java [hunk 3]
```diff
@@ -240,6 +272,29 @@ public void addComposite(Node node) {
     }
 
     /**
+     * Replace a sub-node of this transform node.
+     *
+     * <p>The existing {@link Node} must be a direct child of this {@link Node}. The replacement
+     * node need not be fully specified.
+     */
+    public void replace(Node existing, Node replacement) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java]: Suggest `replaceChild` given the constraints.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java [hunk 4]
```diff
@@ -32,32 +32,42 @@
...
 /**
  * Captures information about a collection of transformations and their
  * associated {@link PValue}s.
  */
 public class TransformHierarchy {
+  private static final Logger LOG = LoggerFactory.getLogger(TransformHierarchy.class);
+
+  private final Pipeline pipeline;
   private final Node root;
+  private final Map<Node, PInput> unexpandedInputs;
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java]: I may have lost the thread of development here, but assuming the replacement is happening after expansion or a serde cycle, we won't actually have an unexpanded input. But we also obviously can't unexpand things without some context. So is it the replacement factory or some such that is responsible for unexpanding inputs in order to apply the replacement?

Or, looking at just this diff, these are just used to have a handle on something to finish specifying?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java (hunk 4 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java]: These are used only in `finishSpecifying`, yes - this probably should be split out, though

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java]: Done.

**Comment by kennknowles** [comment]
LGTM when green. Excited to see this in action.

**Pull request closed without merging**